app/models/user.rb

```diff
@@ -87,11 +88,44 @@ def self.find_for_google_oauth2(access_token, _signed_in_resource = nil)
     user.update!(
       provider: access_token.provider,
       token: access_token.credentials.token,
-      uid: access_token.uid
+      refresh_token: access_token.credentials.refresh_token,
+      uid: access_token.uid,
+      access_expires_at: Time.at(access_token.credentials.expires_at)
     )
     user
   end
 
+  def access_token
+    if !access_expires_at || Time.zone.now > access_expires_at
+      update_access_token
+    else
+      token
+    end
+  end
+
+  def update_access_token
+    url = URI("https://accounts.google.com/o/oauth2/token")
+    refresh_token_params = { 'refresh_token' => refresh_token,
+                              'client_id'     => ENV['GOOGLE_CLIENT_ID'],
+                              'client_secret' => ENV['GOOGLE_CLIENT_SECRET'],
+                              'grant_type'    => 'refresh_token' }
+
+    time_of_request = Time.zone.now
+    # TODO handle a failed request
+    response = Net::HTTP.post_form(url, refresh_token_params)
+    decoded_response = JSON.parse(response.body)
+
+    new_expiration_time =  time_of_request + decoded_response["expires_in"]
+    new_access_token = decoded_response["access_token"]
+
+    self.update!(
```

Comment:
Do we need to confirm that the values for new_expiration_time and new_access_token are non-nil?